tower-circuit-breaker/src/service.rs

```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    fn cause(&self) -> Option<&StdError> {
+        match self {
+            Error::Upstream(ref err) => Some(err),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum State {
+    Closed,
+    Open(Delay),
+}
+
+#[derive(Debug)]
+struct Shared<P, I, T> {
+    state_machine: StateMachine<P, I>,
+    failure_predicate: T,
+}
+
+/// Circuit breaker services, it wraps an inner service instance.
+#[derive(Debug)]
+pub struct CircuitBreakerService<S, P, I, T> {
+    shared: Arc<Shared<P, I, T>>,
+    inner: S,
+    state: State,
+}
+
+impl<S, P, I, T> CircuitBreakerService<S, P, I, T>
+where
+    S: Service,
+    P: FailurePolicy,
+    T: FailurePredicate<S::Error>,
+    I: Instrument,
+{
+    pub fn new(inner: S, failure_policy: P, failure_predicate: T, instrument: I) -> Self {
+        Self {
+            shared: Arc::new(Shared {
+                state_machine: StateMachine::new(failure_policy, instrument),
+                failure_predicate,
+            }),
+            state: State::Closed,
+            inner,
+        }
+    }
+}
+
+impl<S, P, I, T> Service for CircuitBreakerService<S, P, I, T>
+where
+    S: Service,
+    P: FailurePolicy,
+    T: FailurePredicate<S::Error>,
+    I: Instrument,
+{
+    type Request = S::Request;
+    type Response = S::Response;
+    type Error = Error<S::Error>;
+    type Future = ResponseFuture<S::Future, P, I, T>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        if let State::Closed = self.state {
+            if let Some(until) = self.shared.state_machine.is_open() {
+                self.state = State::Open(Delay::new(until))
+            }
+        }
+
+        if let State::Open(ref mut delay) = self.state {
+            let _ = try_ready!(delay.poll().map_err(|_| Error::Rejected));
+        }
+
+        self.state = State::Closed;
+        self.inner.poll_ready().map_err(Error::Upstream)
+    }
+
+    #[inline]
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        ResponseFuture {
+            future: self.inner.call(req),
+            shared: self.shared.clone(),
+            ask: false,
+        }
+    }
+}
+
+pub struct ResponseFuture<F, P, I, T> {
+    future: F,
+    shared: Arc<Shared<P, I, T>>,
+    ask: bool,
+}
+
+impl<F, P, I, T> Future for ResponseFuture<F, P, I, T>
+where
+    F: Future,
+    P: FailurePolicy,
+    T: FailurePredicate<F::Error>,
+    I: Instrument,
+{
+    type Item = F::Item;
+    type Error = Error<F::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        if !self.ask {
+            self.ask = true;
+            if !self.shared.state_machine.is_call_permitted() {
+                return Err(Error::Rejected);
+            }
+        }
+
+        match self.future.poll() {
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Ok(Async::Ready(rep)) => {
+                self.shared.state_machine.on_success();
+                Ok(Async::Ready(rep))
```

Comment:
@olix0r 

#### 1. The response classification

The response classification should be decoupled from the callback mechanism because it could be a generic way to
distinguish request/response errors in the tower stack. It could be bitmask and applies to an error, something like this:

```rust
/// Indicates a generic response error
const Failure: u8 = 0b0000_0001;
/// A response error is seen as a success
const Success: u8 = 0b0000_0010;
/// A request failed but could be retried
const Retriable: u8 = 0b0000_0100;
/// An transport error, the connection should be closed.
const Fatal: u8 = 0b0000_1000;

trait Classify<E> {
  fn classify(err: E) -> u8;
}

```

#### The callbacks

There are a couple problems with this ad-hoc solution.

![Tipical client stask](https://twitter.github.io/finagle/guide/_images/clientstack.svg)

Here is an image with a typical client stack. Thus callbacks need for

* ConnectionPool, to ensure that connection has only one consumer at the moment.
* Request/response timeouts
* Failure accrual, to track request/response statuses.
* Metrics/Traces
* Load balances, to track request/response statuses and timings.

So, there are at least 6 places where callbacks would be used. It looks too complicated to pack/unpack request/response on each step.

I suggest to consider alternative variants:

* Provide the trait for callbacks in tokio. It could be implemented for HTTP(1/2) requests/responses and executed at the transport layer.
* #106, the suggested schema is best fitted to streaming.